activerecord/lib/active_record/secure_token_digest.rb

```diff
@@ -0,0 +1,98 @@
+      #   user.activation_token                                   # => nil
+      #   user.activation_token_digest                            # => nil
+      #
+      #   user.regenerate_activation_token                        # => true
+      #   user.activation_token                                   # => 'mUc3m00RsqyRe'
+      #   user.activation_token_digest                            # => '$2a$10$4LEA7r4YmNHtvlAvHhsYAeZmk/xeUVtMTYqwIvYY76EW5GUqDiP4.'
+      #   user.authenticated?(:activation_token,'notright')       # => false
+      #   user.authenticated?(:activation_token,'mUc3m00RsqyRe')  # => true
+      def has_secure_token_digest(attribute = :token)
+        # Load bcrypt gem only when has_secure_token_digest is used.
+        begin
+          require "bcrypt"
+        rescue LoadError
+          $stderr.puts "You don't have bcrypt installed in your application. Please add it to your Gemfile and run bundle install"
+          raise
+        end
+
+        attr_reader attribute
+
+        include InstanceMethodsOnActivation
+
+        # Encrypts the token into the +_digest+ attribute, only if the
+        # new token is not empty.
+        #
+        #   class User < ActiveRecord::Base
+        #     has_secure_token_digest
+        #   end
+        #
+        #   user = User.new
+        #   user.token = nil
+        #   user.token_digest # => nil
+        #   user.token = 'mUc3m00RsqyRe'
+        #   user.token_digest # => "$2a$10$4LEA7r4YmNHtvlAvHhsYAeZmk/xeUVtMTYqwIvYY76EW5GUqDiP4."
+        define_method("#{attribute}=") do |unencrypted_token|
+          if unencrypted_token.nil?
+            self.send("#{attribute}_digest=", nil)
+          elsif !unencrypted_token.empty?
+            instance_variable_set("@#{attribute}", unencrypted_token)
+            cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST : BCrypt::Engine.cost
+            self.send("#{attribute}_digest=", BCrypt::Password.create(unencrypted_token, cost: cost))
+          end
+        end
+
+        define_method("regenerate_#{attribute}") do
+          self.send("#{attribute}=", self.class.generate_unique_secure_token)
+          self.send("save!")
+        end
+      end
+
+      def generate_unique_secure_token
+        SecureRandom.base58(24)
+      end
+    end
+
+    module InstanceMethodsOnActivation
+      require "active_support/core_ext/securerandom"
+
+      # Returns +true+ if the token is correct, otherwise +false+.
+      #
+      #   class User < ActiveRecord::Base
+      #     has_secure_token_digest
+      #   end
+      #
+      #   user = User.new(name: 'david', token: 'mUc3m00RsqyRe')
+      #   user.save
+      #   user.authenticated?(:token, 'notright')         # => false
+      #   user.authenticated?(:token, 'mUc3m00RsqyRe')    # => true
+      def authenticated?(attribute, unencrypted_token)
```

Comment:
I don't think `authenticated?` makes sense in this context, unless we only want this feature to be used as authentication, if so I against it because we already have a lot of authentication libraries out there way more battle tested and already dealing with a lot of security problems.